src/Internal/ChannelParser.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Amp\Sync\Internal;
+
+use Amp\Parser\Parser;
+use Amp\Serialization\NativeSerializer;
+use Amp\Serialization\SerializationException;
+use Amp\Serialization\Serializer;
+use Amp\Sync\ChannelException;
+use function Amp\Serialization\encodeUnprintableChars;
+
+final class ChannelParser extends Parser
+{
+    private const HEADER_LENGTH = 5;
+
+    /**
+     * @param \Closure(mixed):void $push
+     * @param Serializer $serializer
+     *
+     * @return \Generator
+     *
+     * @throws ChannelException
+     */
+    private static function parser(\Closure $push, Serializer $serializer): \Generator
+    {
+        while (true) {
+            /** @var string $header */
+            $header = yield self::HEADER_LENGTH;
+            $data = \unpack("Cprefix/Llength", $header);
+
+            if ($data["prefix"] !== 0) {
+                $data = $header . yield;
+                throw new ChannelException("Invalid packet received: " . encodeUnprintableChars($data));
+            }
+
+            try {
+                $data = $serializer->unserialize(yield $data["length"]);
+            } catch (SerializationException $exception) {
+                throw new ChannelException(
+                    "Failed to unserialize the channel data: " . $exception->getMessage(),
+                    0,
+                    $exception,
+                );
+            }
+
+            try {
+                $push($data);
+            } catch (\Throwable $exception) {
+                throw new ChannelException(
+                    "Invoking the parser callback failed: " . $exception->getMessage(),
+                    0,
+                    $exception,
+                );
+            }
+        }
+    }
+
+    private Serializer $serializer;
+
+    /**
+     * @param \Closure(mixed):void $closure Closure invoked when data is parsed.
+     * @param Serializer|null $serializer
+     */
+    public function __construct(\Closure $closure, ?Serializer $serializer = null)
+    {
+        $this->serializer = $serializer ?? new NativeSerializer;
+        parent::__construct(self::parser($closure, $this->serializer));
+    }
+
+    /**
+     * @param mixed $data Data to encode to send over a channel.
+     *
+     * @return string Encoded data that can be parsed by this class.
+     *
+     * @throws ChannelException
+     */
+    public function encode(mixed $data): string
```

Comment:
Should a `Parser` encode data? :stuck_out_tongue_winking_eye:

Reply:
Naming is probably bad… it's encoding what the parser expects. It's internal, so I don't much care.

Reply:
Same ;-9